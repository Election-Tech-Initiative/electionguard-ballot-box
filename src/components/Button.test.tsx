import React from 'react'
import { fireEvent, render } from '@testing-library/react'

import Button, { DecoyButton } from './Button'

const createTouchStartEventProperties = (x: number, y: number) => {
  return { touches: [{ clientX: x, clientY: y }] }
}

const createTouchEndEventProperties = (x: number, y: number) => {
  return { changedTouches: [{ clientX: x, clientY: y }] }
}

const onPress = jest.fn()

it('renders Button', () => {
  const { container } = render(<Button onPress={onPress}>foo</Button>)
  expect(container.firstChild).toMatchSnapshot()
})

it('renders primary Button', () => {
  const { container } = render(
    <Button onPress={onPress} primary>
      Primary
    </Button>
  )
  expect(container.firstChild).toMatchSnapshot()
})

it('renders danger Button', () => {
  const { container } = render(
    <Button onPress={onPress} danger>
      Danger!
    </Button>
  )
  expect(container.firstChild).toMatchSnapshot()
})

it('renders big Button', () => {
  const { container } = render(
    <Button onPress={onPress} big>
      I’m a big button!
    </Button>
  )
  expect(container.firstChild).toMatchSnapshot()
})

it('renders small Button', () => {
  const { container } = render(
    <Button onPress={onPress} small>
      I’m a small button!
    </Button>
  )
  expect(container.firstChild).toMatchSnapshot()
})

it('renders DecoyButton', () => {
  const { container } = render(<DecoyButton>DecoyButton</DecoyButton>)
  expect(container.firstChild).toMatchSnapshot()
})

it('works properly with clicks and touches', () => {
  // eslint-disable-next-line no-shadow
  const onPress = jest.fn()
  const { getByText } = render(<Button onPress={onPress}>Test Button</Button>)
  const button = getByText('Test Button')

  fireEvent.click(button)
  expect(onPress).toHaveBeenCalledTimes(1)

  // TouchEnd close to TouchStart calls onPress.
  fireEvent.touchStart(button, createTouchStartEventProperties(100, 100))
  fireEvent.touchEnd(button, createTouchEndEventProperties(110, 95))
  expect(onPress).toHaveBeenCalledTimes(2)

  // Using preventDefault() with touch prevents the click, so no need to test click de-duping.

  // TouchStart w/o TouchEnd does not call onPress.
  fireEvent.touchStart(button, createTouchStartEventProperties(100, 100))
  expect(onPress).toHaveBeenCalledTimes(2)

  // TouchEnd too far from TouchStart does not call onPress.
  fireEvent.touchEnd(button, createTouchEndEventProperties(131, 95))
  expect(onPress).toHaveBeenCalledTimes(2)

  // Keyboard (also Accessible Controller) fire click event which calls onPress.
  fireEvent.click(button)
  expect(onPress).toHaveBeenCalledTimes(3)
})
