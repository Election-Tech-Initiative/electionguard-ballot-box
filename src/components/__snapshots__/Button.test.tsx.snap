// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Button 1`] = `
<button
  class="sc-bwzfXH gBzjSr"
  type="button"
>
  foo
</button>
`;

exports[`renders DecoyButton 1`] = `
<div
  class="sc-bdVaJa fUKbkn"
>
  DecoyButton
</div>
`;

exports[`renders big Button 1`] = `
<button
  class="sc-bwzfXH bbSMPw"
  type="button"
>
  I’m a big button!
</button>
`;

exports[`renders danger Button 1`] = `
<button
  class="sc-bwzfXH kKlRTC"
  type="button"
>
  Danger!
</button>
`;

exports[`renders primary Button 1`] = `
<button
  class="sc-bwzfXH gvKole"
  type="button"
>
  Primary
</button>
`;

exports[`renders small Button 1`] = `
<button
  class="sc-bwzfXH hyvTSe"
  type="button"
>
  I’m a small button!
</button>
`;
